### PR TITLE
[WAF] Add data source huaweicloud_waf_dedicated_instances to query WAF dedicated engines.

### LINKED ISSUE #1355: [WAF] Add data sources and resources for dedicated engine.
### Current Terraform Version

```
1.27.0
```

This is an internal requirement as follows:
- Add the data source to query WAF policies, source name is `huaweicloud_waf_policies`.
- Manage domain names in Premium-waf, resource name is `huaweicloud_waf_dedicated_domain`.
  Functions include: creating a domain names, modifying protection status, modifying protection mode, and deleting domain names.
- Manage engine in Premium-waf, resource name is `huaweicloud_waf_dedicated_engine`.
  Functions include: creating a dedicated engine,deleting dedicated,querying dedicated engine by id.
- We also needs to query the dedicated engine infomation, the data source name is `huaweicloud_waf_dedicated_engines` and `huaweicloud_waf_dedicated_engine`.

### PR DESCRIPTION
**Pull request by chengxiangdong**
**What this PR does / why we need it**:

- Add the data source to query a WAF dedicated engines, data source name is `huaweicloud_waf_dedicated_engines `.

**Which issue this PR fixes**:
*(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*
fixes #1355

**Special notes for your reviewer**:

NONE

**Release note**:

```release-note
NONE
```

## PR Checklist

* [x] Tests added/passed.
* [x] Documentation updated.
* [x] Schema updated.

## Acceptance Steps Performed

```
make testacc TEST='./huaweicloud/services/acceptance/waf' TESTARGS='-run TestAccDataSourceWafDedicatedInstancesV1_basic'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./huaweicloud/services/acceptance/waf -v -run TestAccDataSourceWafDedicatedInstancesV1_basic -timeout 360m -parallel 4
=== RUN   TestAccDataSourceWafDedicatedInstancesV1_basic
--- PASS: TestAccDataSourceWafDedicatedInstancesV1_basic (362.14s)
PASS
ok      github.com/huaweicloud/terraform-provider-huaweicloud/huaweicloud/services/acceptance/waf       362.278s


```

### CODE REVIEW

File: docs/data-sources/waf_dedicated_instances.md [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+```
+
+## Argument Reference
+
+The following arguments are supported:
+
+* `region` - (Optional, String) The region in which to query the WAF dedicated instance.
+  If omitted, the provider-level region will be used. 
+
+* `instance_id` - (Optional, String) The id of WAF dedicated instance. Only one of `instance_id` and `name` can be set.
```
Comment (ShiChangkuo) [inline docs/data-sources/waf_dedicated_instances.md]: We can remove this restriction by check the name after get instance by ID

File: huaweicloud/services/waf/data_source_huaweicloud_waf_dedicated_instances.go [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+
+func dataSourceWafDedicatedInstanceV1Read(d *schema.ResourceData, meta interface{}) error {
+	config := meta.(*config.Config)
+	client, err := config.WafDedicatedV1Client(config.GetRegion(d))
+	if err != nil {
+		return fmtp.Errorf("error creating HuaweiCloud WAF dedicated client: %s", err)
+	}
+
+	var items []instances.DedicatedInstance
+	if len(d.Get("instance_id").(string)) > 0 {
```
Comment (ShiChangkuo) [inline huaweicloud/services/waf/data_source_huaweicloud_waf_dedicated_instances.go]: please use d.GetOk instead

Comment (chengxiangdong) [reply huaweicloud/services/waf/data_source_huaweicloud_waf_dedicated_instances.go]: fixed

File: docs/data-sources/waf_dedicated_instances.md (hunk 1 above)
Comment (chengxiangdong) [reply docs/data-sources/waf_dedicated_instances.md]: `instance_id` has been changed to `id`

File: docs/data-sources/waf_dedicated_instances.md [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+
+## Attributes Reference
+
+The following attributes are exported:
+
+* `instances` -  A list of WAF dedicated instances.
+
+The `instances` block supports:
+
+* `name` - The name of WAF dedicated instance.
```
Comment (ShiChangkuo) [inline docs/data-sources/waf_dedicated_instances.md:40]: instances/*/id is missing

File: huaweicloud/services/waf/data_source_huaweicloud_waf_dedicated_instances.go [hunk 4]
```diff
@@ -0,0 +1,174 @@
...
+	if err != nil {
+		return fmtp.Errorf("error creating HuaweiCloud WAF dedicated client: %s", err)
+	}
+
+	instanceId, hasId := d.GetOk("id")
+	var items []instances.DedicatedInstance
+	if hasId {
+		ins, err := instances.GetInstance(client, instanceId.(string))
+		if err != nil {
+			return common.CheckDeleted(d, err, "Error obtain WAF dedicated instance by id information.")
```
Comment (ShiChangkuo) [inline huaweicloud/services/waf/data_source_huaweicloud_waf_dedicated_instances.go]: just return the error if the id does not exist

File: docs/data-sources/waf_dedicated_instances.md (hunk 3 above)
Comment (chengxiangdong) [reply docs/data-sources/waf_dedicated_instances.md:40]: added

**Pull request merged**